src/core/thread/link_metrics.cpp

```diff
@@ -0,0 +1,391 @@
+    uint16_t               startOffset      = aMessage.GetLength();
+    uint16_t               readOffset;
+    uint16_t               endOffset;
+    otLinkMetrics          linkMetrics;
+
+    memset(&linkMetrics, 0, sizeof(linkMetrics));
+
+    SuccessOrExit(error = Tlv::FindTlvOffset(aRequestMessage, Mle::Tlv::Type::kLinkMetricsQuery, readOffset));
+    VerifyOrExit(aRequestMessage.Read(readOffset, sizeof(tlv), &tlv) == sizeof(tlv), error = OT_ERROR_PARSE);
+
+    readOffset += sizeof(Tlv);
+    endOffset = readOffset + tlv.GetLength();
+
+    while (readOffset < endOffset)
+    {
+        VerifyOrExit(aRequestMessage.Read(readOffset, sizeof(tlv), &tlv) == sizeof(tlv), error = OT_ERROR_PARSE);
+
+        switch (tlv.GetType())
+        {
+        case kLinkMetricsQueryId:
+            SuccessOrExit(error = Tlv::ReadUint8Tlv(aRequestMessage, readOffset, queryId));
+            hasQueryId = true;
+            break;
+
+        case kLinkMetricsQueryOptions:
+        {
+            for (uint16_t offset = readOffset + sizeof(tlv); offset < readOffset + sizeof(tlv) + tlv.GetLength();
+                 offset += sizeof(LinkMetricsTypeIdFlags))
+            {
+                LinkMetricsTypeIdFlags typeIdFlags;
+                VerifyOrExit(aRequestMessage.Read(readOffset + sizeof(tlv), sizeof(typeIdFlags), &typeIdFlags) ==
+                                 sizeof(typeIdFlags),
+                             error = OT_ERROR_PARSE);
+                switch (typeIdFlags.GetMetricEnum())
+                {
+                case OT_LINK_METRICS_PDU_COUNT:
+                    VerifyOrExit(typeIdFlags.IsLengthFlagSet() || !linkMetrics.mPduCount, error = OT_ERROR_PARSE);
+                    linkMetrics.mPduCount = true;
+                    break;
+
+                case OT_LINK_METRICS_LQI:
+                    VerifyOrExit(!typeIdFlags.IsLengthFlagSet() || !linkMetrics.mLqi, error = OT_ERROR_PARSE);
+                    linkMetrics.mLqi = true;
+                    break;
+
+                case OT_LINK_METRICS_MARGIN:
+                    VerifyOrExit(!typeIdFlags.IsLengthFlagSet() || !linkMetrics.mLinkMargin, error = OT_ERROR_PARSE);
+                    linkMetrics.mLinkMargin = true;
+                    break;
+
+                case OT_LINK_METRICS_RSSI:
+                    VerifyOrExit(!typeIdFlags.IsLengthFlagSet() || !linkMetrics.mRssi, error = OT_ERROR_PARSE);
+                    linkMetrics.mRssi = true;
+                    break;
+
+                default:
+                    break;
+                }
+
+                typeIdFlagsArray[typeIdFlagsCount++] = typeIdFlags;
+            }
+            break;
+        }
+
+        default:
+            break;
+        }
+        readOffset += sizeof(Tlv) + tlv.GetLength();
+    }
+
+    VerifyOrExit(hasQueryId, error = OT_ERROR_PARSE);
+
+    // Link Metrics Report TLV
+    tlv.SetType(Mle::Tlv::kLinkMetricsReport);
+    SuccessOrExit(error = aMessage.Append(&tlv, sizeof(tlv)));
```

Comment:
My thought is that this function `AppendLinkMetricsReport` would handle both cases for Single Probe and Forward tracking series. So the Report TLV header would be the same for both cases. So I think we can put it here so that it works for both cases. Thoughts?

Reply:
SGTM.